testing/src/test/java/io/stargate/it/storage/StargateContainer.java

```diff
@@ -0,0 +1,538 @@
+ * extensions are activated in the right order.
+ *
+ * <p>Note: this extension does not support concurrent test execution.
+ *
+ * @see StargateSpec
+ * @see StargateParameters
+ */
+public class StargateContainer
+    implements BeforeAllCallback,
+        AfterAllCallback,
+        BeforeEachCallback,
+        AfterEachCallback,
+        ParameterResolver {
+  private static final Logger LOG = LoggerFactory.getLogger(StargateContainer.class);
+
+  private static final File LIB_DIR = new File(System.getProperty("stargate.libdir"));
+  private static final int PROCESS_WAIT_MINUTES =
+      Integer.getInteger("stargate.test.process.wait.timeout.minutes", 10);
+
+  private static final AtomicBoolean suiteExecuting = new AtomicBoolean();
+  private static final AtomicBoolean testExecuting = new AtomicBoolean();
+
+  // the first 10 addresses are reserved for storage nodes
+  private static final AtomicInteger stargateAddressStart = new AtomicInteger(11);
+  private static final AtomicInteger stargateInstanceSeq = new AtomicInteger();
+
+  private static Container container;
+
+  static {
+    Thread containerShutdown =
+        new Thread("stargate-container-shutdown") {
+          @Override
+          public void run() {
+            try {
+              Container c = container;
+              if (c != null) {
+                c.stop();
+              }
+            } catch (Exception e) {
+              LOG.warn("Exception during global container shutdown: {}", e.getMessage(), e);
+            }
+          }
+        };
+
+    Runtime.getRuntime().addShutdownHook(containerShutdown);
+  }
+
+  private static Container container() {
+    if (container == null) {
+      throw new IllegalStateException("Stargate container has not been configured");
+    }
+
+    return container;
+  }
+
+  private static void enter(AtomicBoolean flag) {
+    if (!flag.compareAndSet(false, true)) {
+      throw new IllegalStateException(
+          String.format(
+              "Concurrent execution with %s is not supported",
+              StargateContainer.class.getSimpleName()));
+    }
+  }
+
+  private static void leave(AtomicBoolean flag) {
+    flag.compareAndSet(true, false);
+  }
+
+  private static StargateParameters parameters(StargateSpec spec, ExtensionContext context)
+      throws Exception {
+    Builder builder = StargateParameters.builder();
+
+    String customizer = spec.parametersCustomizer().trim();
+    if (!customizer.isEmpty()) {
+      Object testInstance = context.getTestInstance().orElse(null);
+      Class<?> testClass = context.getRequiredTestClass();
+      Method method = testClass.getMethod(customizer, Builder.class);
+      method.invoke(testInstance, builder);
+    }
+
+    return builder.build();
+  }
+
+  private static synchronized void start(ExtensionContext context) throws Exception {
+    ClusterConnectionInfo backend = ExternalStorage.connectionInfo();
```

Comment:
Can we use JUnit's parameter injection instead of relying on static state? It would be nice to say "use the ExternalStorage that is active for this test".

Reply:
I do not think parameter injection works between extensions. I use the `ExtensionContex.Store` for that initially, but the code looked actually messier that static accessors :) 

I'll try something else presently :thinking:

Reply:
It's kind of disappointing if there's no easy way to share information between extensions.

For the Java driver we passed the actual rule instance to the dependent rule ([example](https://github.com/datastax/java-driver/blob/4.x/integration-tests/src/test/java/com/datastax/oss/driver/core/cql/SimpleStatementCcmIT.java#L50-L58)). Maybe we could do something similar with JUnit 5 extensions, but you'd have to declare every rule programmatically as a static or instance field (no more `@ExtendWith`).

Reply:
I'll have another try with `ExtensionContex.Store`. We could put the key name into the "spec" annotation in case we need to support multiple storage clusters in the same test env. :thinking:

Reply:
Storage into objects are passed through JUnit contexts now.